onnxruntime/core/providers/cuda/math/softmax_warpwise_impl.cuh

```diff
@@ -163,5 +163,71 @@ __global__ void softmax_warp_forward(output_t* dst, const input_t* src, int batc
   }
 }
 
+
+// softmax_warp_forward use register to store data and the dtype is fp32, which will result resource oversubscription and thus hurt kernel performance.
+// softmax_warp_forward_resource_efficient is implemneted to solve the issues.
+template <typename input_t, typename output_t, typename acc_t, int log2_elements, bool is_log_softmax>
+__global__ void softmax_warp_forward_resource_efficient(output_t* dst, const input_t* src, int batch_size, int stride, int element_count) {
+  // 1 cuda block only processes one row and contains 1 cuda warp only.
+  constexpr int next_power_of_two = 1 << log2_elements;
+  constexpr int WARP_SIZE = (next_power_of_two < GPU_WARP_SIZE) ? next_power_of_two : GPU_WARP_SIZE;
+  constexpr int WARP_ITERATIONS = next_power_of_two / WARP_SIZE;
+
+  int local_idx = threadIdx.x;
+  src +=  blockIdx.x * stride + local_idx;
+  dst +=  blockIdx.x * stride + local_idx;
+  extern __shared__ unsigned char smem[];
+  input_t (&elements)[WARP_ITERATIONS][WARP_SIZE] = *reinterpret_cast<input_t (*)[WARP_ITERATIONS][WARP_SIZE]>(smem);
+#pragma unroll
+  for (int it = 0; it < WARP_ITERATIONS; ++it) {
+    int element_index = local_idx + it * WARP_SIZE;
+    if (element_index < element_count) {
+      elements[it][local_idx] = src[it * WARP_SIZE];
+    } else {
+      elements[it][local_idx] = -std::numeric_limits<input_t>::infinity();
+    }
+  }
+  // compute max_value
+  input_t max_value = elements[0][local_idx];
+#pragma unroll
+  for (int it = 1; it < WARP_ITERATIONS; ++it) {
+    max_value = (max_value > elements[it][local_idx]) ? max_value : elements[it][local_idx];
+  }
+  warp_reduce<input_t, 1, WARP_SIZE, Max>(&max_value);
+  // compute sum
+  acc_t sum{0.0f};
+  // #pragma unroll
```

Comment:
interesting findings. :)

Reply:
if so, do you think we should also remove it for `softmax_warp_forward`?

Reply:
it depends on how many iteration will be unrolled, when iter number is small, the icache miss is ok, so we need a threshold